### PR TITLE
[2/Z] Add feature that allows user to limit number of comments shown.

### PR DESCRIPTION
**Pull request by nikhitashanker**
In this change, we create an input field that the user can use to limit the number of comments shown.

### CODE REVIEW

**Comment by nikhitashanker** [comment]
I will rebase after [1/Z] is merged,

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -53,16 +67,71 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    // Get input from the form.
+    String numberOfCommentsString = request.getParameter("number-of-comments");
+
+    // Convert input to an int.
+    int numberOfComments;
+    try {
+      numberOfComments = Integer.parseInt(numberOfCommentsString);
+    } catch (NumberFormatException e) {
+      System.err.println("Cannot convert to int: " + numberOfCommentsString);
+      return -1;
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:94]: It is not recommended to use a illegal number to present failure.

I think you can either:
  1, Throw the Exception and catch it outside.
  2, Or make the function to return Optional<Integer>. So you can return Optional.empty() when you see failure.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -27,12 +34,18 @@
...
+    // Get input from the form.
+    int numberOfComments = getNumberOfComments(request);
+    if (numberOfComments == -1) {
+      response.setContentType("text/html");
+      response.getWriter().println("Please enter an integer greater than 1");
+      return;
+    }
+
     response.setContentType("application/json;");
+    List<Comment> comments = getCommentsFromDataStore(numberOfComments);
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: Inline this 'comments'.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -53,16 +67,71 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+  private static int getNumberOfComments(HttpServletRequest request) {
+    // Get input from the form.
+    String numberOfCommentsString = request.getParameter("number-of-comments");
+
+    // Convert input to an int.
+    int numberOfComments;
+    try {
+      numberOfComments = Integer.parseInt(numberOfCommentsString);
+    } catch (NumberFormatException e) {
+      System.err.println("Cannot convert to int: " + numberOfCommentsString);
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: Can you use log here? Maybe comfirm with the support engineers if System.println is the recommend way to log for your project?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -53,16 +67,71 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+  private static String getParameter(HttpServletRequest request, String name, String defaultValue) {
     String value = request.getParameter(name);
     return (value == null || value.isEmpty()) ? defaultValue : value;
   }
+
+  /** Returns number of comments entered by the user or -1 if choice is invalid */
+  private static int getNumberOfComments(HttpServletRequest request) {
+    // Get input from the form.
+    String numberOfCommentsString = request.getParameter("number-of-comments");
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:86]: use a "private static final String" const for the "number-of-comments"

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: You can use String.format("Cannot convert to int: %s", numberOfCommentsString) here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -53,16 +67,71 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    try {
+      numberOfComments = Integer.parseInt(numberOfCommentsString);
+    } catch (NumberFormatException e) {
+      System.err.println("Cannot convert to int: " + numberOfCommentsString);
+      return -1;
+    }
+
+    // Check that the input is greater than 0.
+    if (numberOfComments < 0) {
+      System.err.println("Number of comments specified is out of range: " + numberOfCommentsString);
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:99]: ditto

**Comment by nikhitashanker** [comment]
I rebased after merging the datastore pull request to master [1/Z] and pushed to this branch.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:99]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:86]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: The support engineers said that we can use a logging framework like Log4j but for this project System.err.println is sufficient.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:94]: Done. 1st option taken - throw the exception and catch it outside (caught in the doGet Method and thrown in the getNumberOfComments method)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -33,21 +34,23 @@
...
+      System.err.println(e.getMessage());
+      response.setContentType("text/html");
+      response.getWriter().println("Please enter an integer value greater than 1");
+      return;
     }
 
-    // Send the JSON as the response.
     response.setContentType("application/json;");
-    response.getWriter().println(convertToJson(comments));
+    response.getWriter().println(convertToJson(getCommentsFromDataStore(numberOfComments)));
```
Comment (hlz03) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:53]: You can put these in try block.

Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -33,21 +34,22 @@
 /** Servlet that handles comments data. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  public static final String PARAMETER = "number-of-comments";
```
Comment (liangqigong) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:37]: Change the name since 'PARAMETER' is very vague.

Comment (nikhitashanker) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:37]: Done.

**Pull request merged**